src/lambda_toolshed/papillon.cljc

```diff
@@ -68,46 +65,69 @@
 
 (defn- move
   "Move to the next interceptor in the interceptor chain within `ctx`.  Returns
-  a tuple of the resulting context, the current interceptor and a descriptive
-  tag of the move operation.  Returns nil if the chain has been consumed."
-  [{::keys [queue stack stage trace] :as ctx}]
+  a three-tuple of the resulting context, the current interceptor and the
+  resulting stage keyword.  Returns nil if the chain has been consumed."
+  [{::keys [queue stack stage error handled?] :as ctx}]
   (case stage
     :enter (if-let [ix (peek queue)]
-             (let [tag [(identify ix) stage]
-                   ctx (cond-> ctx
-                         true (update ::queue pop)
-                         true (update ::stack conj ix)
-                         trace (update ::trace conj tag))]
-               [ctx ix tag])
-             (-> ctx (assoc ::stage :leave) move))
-    (:error :leave) (if-let [ix (peek stack)]
-                      (let [tag [(identify ix) stage]
-                            ctx (cond-> ctx
-                                  true (update ::stack pop)
-                                  trace (update ::trace conj tag))]
-                        [ctx ix tag])
-                      nil)))
+             (let [ctx (-> ctx
+                           (update ::queue pop)
+                           (update ::stack conj ix))]
+               [ctx ix :enter])
+             (recur (assoc ctx ::stage (if error :error :leave))))
+    :leave (when-let [ix (peek stack)]
+             (let [ctx (-> ctx
+                           (assoc ::stage :final))]
+               [ctx ix :leave]))
+    :error (when-let [ix (peek stack)]
+             (let [ctx (-> ctx
+                           (assoc ::stage :final)
+                           (assoc ::handled? true))]
+               [ctx ix :error]))
+    :final (if (or (nil? error) handled?)
+             (if-let [ix (peek stack)]
+               (let [ctx (-> ctx
+                             (update ::stack pop)
+                             (dissoc ::handled?)
+                             (assoc ::stage (if error :error :leave)))]
+                 [ctx ix :final])
+               nil)
+             (recur (assoc ctx ::stage :error)))))
 
 (defn- evaluate
-  "Evaluate the interceptor `ix` with the context `ctx`."
-  [ix {::keys [stage] :as ctx}]
-  (if-let [f (ix stage)]
-    (try (f ctx) (catch #?(:clj Throwable :cljs :default) e e))
-    ctx))
+  "Evaluate the next operation of the given context `ctx`. Returns a three-tuple
+  of the context prior to interceptor execution, the emerge-able result of
+  interceptor execution and a tag documenting the execution."
+  [ctx]
+  (when-let [[ctx ix stage] (move ctx)]
+    (let [tag [(identify ix) stage]
+          ctx (update ctx ::trace (fn [t] (when t (conj t tag))))
+          f (or (ix stage) identity)
```

Comment:
A question as I look at this fresh regarding the trace...

Do we want to move the `f` resolving up in the let, and check if it is not identity before adding a trace, then we see the trace of functions defined?

```clojure
(let [tag [(identify ix) stage]
      f (or (ix stage) identity)
      ctx (update ctx ::trace (fn [t] (when (and t
                                                 (not (= f identity)))
                                        (conj t tag))))
```

not sure if that would make the trace "nicer" or not.

Then the trace is the trace of what was defined on the interceptors vs "inferred"

Reply:
I wondered about this too and elected to leave it alone.  The pros of leaving the trace "complete" is that it additionally documents what "could have been exploited" -so a bit of a training aid.  The cons are noise when you only care which of *your* functions were executed.